README

                                   OutBoard
                          By Richard F. Feuerriegel
                      ==================================

Current version: 2.2.6                            Release Date: 2012-04-25


Official Web site
-----------------
	http://outboard.sourceforge.net/


Brief Overview
--------------
This program emulates an office "in/out" board using PHP and MySQL. The
board is displayed in a browser window that updates on a regular basis.
There is an administrative area, and a timeclock reports section.


Server/Client Requirements
--------------------------
	+ (Client) JavaScript enabled in the web browser
	+ Apache 1.3.X or 2.X (http://www.apache.org)
	+ PHP 5.X Apache module (http://www.php.net)
	+ MySQL 5.X (http://www.mysql.com)
	+ Basic understanding of how to use the above software. :-)


Major Features
--------------
Below are the major, new features for each version. More detailed
information is in the CHANGES file.

	New for version 2.2.6 (2012-04-25):
	-----------------------------------
	+ Compatible with MySQL 5.1+ after timestamp(14) issue resolved. 

	New for version 2.2.4 (2007-02-16):
	-----------------------------------
	+ Customizable session cookie timeout

	New for version 2.2.3 (2006-12-06):
	-----------------------------------
	+ Zebra striping of user rows for alternating colors

	New for version 2.2.0 (2005-03-04):
	---------------------------------
	+ Configurable setting to disallow a user to change someone
	  else's dot and remarks.

	New for version 2.1.0 (2005-02-19):
	---------------------------------
	+ The board now scrolls to the correct editing position
	+ An optional link can be created via the config file to 
	  open a small window for arbitrary uses, such as employee 
	  schedules.

	New for version 2.0.0 (2005-02-18):
	---------------------------------
	+ Timeclock reports for users and administrators
	+ Automatic moving of dots to "out" if users are idle
	+ Major source code refactoring on many files
	+ New classes and more object-oriented programming
	+ OB header reprints after set number of rows
	+ Passwords are now encrypted in the database.

	New for version 1.4.0 (2002-04-15):
	---------------------------------
	+ Basic Authentication is used automatically if available
	+ Multiple users can have Administration privileges
	+ Updated character widths file
	+ Minor bug fixes

	New for version 1.3.0 (2001-06-12):
	---------------------------------
	+ Automatic installation of necessary tables
	+ Integrated user administration interface
	+ Integrated authentication system (.htaccess optional)
	+ Updated character widths file
	+ Minor bug fixes 

	New for version 1.2.0 (2000-11-13):
	---------------------------------
	+ Simple interface that emulates a real "in/out" board
	+ Accidental write protection via a "change" button
	+ Ten "return by" times (8:00AM to 5:00PM)
	+ Separate In and Out columns
	+ Column for Remarks, with automatic "anti-wrapping" 
	+ Automatic page refresh with daytime and nighttime rates
	+ Javascript 1.2 compliant (and required)
	+ Netscape 4 and Internet Explorer 5 tested
	+ Hovering over a dot shows when it was last changed
	+ Automatic logging of changes made to the board


	
                   -------------------------------
                   -- Installation Instructions --
                   -------------------------------

-----------------------------------------------------------------------
IMPORTANT: If you already have the OutBoard installed, please follow the
UPGRADE instructions instead of these. 

------> Backup your config.php file! <------


STEP 1.  Unpack the outboard-X.X.tar.gz file, and change the directory 
name, (or make a symbolic link) to "outboard" off your web root:

  cd WEB_DOCUMENT_ROOT_DIRECTORY      (example: cd /opt/apache/htdocs)
  mv outboard-X.X.tar.gz  .
  tar -xvzf outboard-X.X.tar.gz       (X.X will be the version number)

OPTIONAL:
  mv outboard-X.X outboard

The URL to it should be something like this:

	http://www.someplace.com/outboard-X.X

If you changed the name of the directory to outboard, the URL will be:

	http://www.someplace.com/outboard

You will  need to make sure that the user which the web server runs as
has access to read the files in the outboard directory. In many cases,
the server runs as "apache" or "nobody", so you whould perform the 
folling commands:

	chgrp -R apache outboard-X.X
	chmod -R 755 outboard-X.X


-----------------------------------------------------------------------
STEP 2. Edit the config/config.php file to match your configuration. 

You will need to at least check/set the database server name, database 
name, username, and the password for the database user. The OutBoard 
database user must have full permission to that database. (Insert, 
Delete, Select, etc.)


-----------------------------------------------------------------------
STEP 3. Create the MySQL Database 

In the config.php file, which you should have edited in STEP 2, there
is a variable that controls to which database the OutBoard will talk.
By default, this is set to "outboard". You can keep this, in which
case you will need to make a new database by that name (or you can 
choose an existing database.) It is recommended that you use a NEW
database, and name it "outboard".

To make a new database in which to store the tables, run the following 
command:

	mysqladmin -u USERNAME --password=PASSWORD create outboard

Replace USERNAME and PASSWORD with the apropriate information for your
MySQL installation. Please refer to your MySQL documentation for
further information about how to create a database, and set permissions
on that database. 

If you have an existing database that you want to use (this is not 
recommended), make sure its information is in the config.php file.
   

-----------------------------------------------------------------------
STEP 4. Initiate the automatic table installation

At this point, the OutBoard will be ready to install the tables that
it needs in the chosen database. You just need to point your browser
to the board's URL (MAKE SURE POP-UPS ARE ENABLED!):

	http://www.somewhere.com/outboard   (or outboard-X.X [Step 0])

A new window will be created (if you have JavaScript enabled), and
you will be prompted to check the configuration. Enter a new admin
username, name, and password, and then click the Install button.

This step will FAIL if the database information at the top of the
config.php file is not correct. Errors at this point are most likely
with MySQL permissions. Check your MySQL documentation first if you
have problems at this stage. Also, read about the MySQL reload command.

As an extra security measure, RENAME the include/install.php script. 
It should not be able to run once you change the installtables 
configuration setting to false, but it's better to be safe than sorry.

-----------------------------------------------------------------------
STEP 4a) OPTIONAL -- Install the Basic Authentication files:

By default the OutBoard will use its own internal authentication 
system. Optionally, you can have it use the Basic Authentication
information provided by an .htaccess file. 

You can create an .htaccess file (in the Outboard directory) with the 
following lines:

AuthUserFile /path_to_apache_password_file/passwd
AuthName "Outboard"
AuthType Basic
require valid-user

You can also use "require user username1 username2" as the last line. 
If you already have a usable Apache password file, point to it on
the AuthUserFile line. Refer to your Apache documentation on how to
create a password file.


-----------------------------------------------------------------------
STEP 5. Access the outboard with your web browser. :-)  See STEP 1.

You will first need to login as the admin user using your chosen 
password, and create new users. 


Notes
-----

* READONLY users can login to the board, but their name will not
show up in the list, nor will they be able to change anything.

* Internal passwords (in the DB) are not used if basic authentication
is in use.

* You might have to turn off magic quotes in the php.ini file if you
get "backslash multiplication" problems. 


History
-------
Version 2.2.6                                  Release Date: 2012-04-25
Version 2.2.5                                  Release Date: 2009-05-07
Version 2.2.4                                  Release Date: 2007-02-16
Version 2.2.3                                  Release Date: 2006-12-06
Version 2.2.2                                  Release Date: 2005-10-25
Version 2.2.1                                  Release Date: 2005-03-06
Version 2.2.0                                  Release Date: 2005-03-04
Version 2.1.0                                  Release Date: 2005-02-19
Version 2.0.0                                  Release Date: 2005-02-18
Version 1.4.0                                  Release Date: 2002-04-15
Version 1.3.0                                  Release Date: 2001-06-12
First released version: 1.2                    Release Date: 2000-11-13



CHANGES
2012-04-25
	+	Changed include/install.php to remove some deprecated PHP MySQL functions
		and use the newer versions:
			* Added mysql_select_db() call near beginning 
			* Changed mysql_db_query() to mysql_query()
	+	Changed deprecated ereg() functions to preg_match() in several files.
	+	Added a way to set the program's timezone in the config.php file.

2012-03-28
	+	Added three alternate files for UNSUPPORTED PostgreSQL compatibility.
		See the README-postgreSQL for more information.

2012-02-28
	+	Changed lib/OutboardDatabase.php so that MySQL connections are not
		persistent. This should help with large installations of the OB.
		Thank you to "sirglandon" on the forums for bringing this to my 
		attention.
	+	Changed "timestamp(14)" to "timestamp" for compatibility with later
		versions of MySQL.

2009-08-10
	+	Changed SQL code in include/create_tables.php so that last_change 
		can be null. It was NOT NULL before which caused the program to fail
		sometimes because it could not insert the initial administrative user.

2009-05-07
	+	Released version 2.2.5

2007-11-14
	+	Fixed a bug with Remarks that caused the clickable link to not
		be displayed in certain circumstances if the user entered only
		a space for the remark and saved it.

2007-05-29
	+	Fixed a CSS problem (removed a misplaced semicolon.)
	+	Moved the urlencode() of username out of the loop in outboard.php to
		fix a bug that caused the hilighting of the user's row not to happen
		in certain cases.
	
2007-02-16
	+	Released version 2.2.4
	+	Added customizable cookie timeout for the session
	+	Fixed a bug in the zebra striping code that looked for a specific user.

2006-12-16
	+	Released version 2.2.3

2006-06-15
	+	Added a "Zebra striping" option so that you can have alternating
		color on the user rows.
	+	Graphics changed to transparent PNGs to support zebra striping.
	+	Remarks bugfix where double-quotes would cause problems.


2005-10-25
	+	Released version 2.2.2.

2005-09-09
	+	Changed include/create_tables.php to be compatible with MySQL 4.1 
		and higher. The auto_increment rowid column needed the "default '0' 
		to be removed.
		
2005-08-29
	+	Changed loginscreen.php so that the BasicAuthInUse error message is
		not "you are not a member of this board" but just "you are not logged
		in." The old message was confusing.

2005-08-17
	+	Added temp_dir configuration variable to set the temporary directory
		for writing the report HTML files. This will allow people on Windows
		systems to change it easily from /tmp/ to C:\something.
	+	Fixed a bug that was putting an extra backslash (\) in user data. It
		turned out that addslashes() was being called twice on the same data
		before putting it in the DB.

2005-08-10
	+	Fixed a bug in the automatic logout routine that caused it to run, 
		even if the configuration value was set to zero (0).

2005-07-26
	+	Added checks in include/install.php and lib/OutboardDatabase.php 
		for the existance of mysql functions to aid in the debugging of 
		installation and running issues.

2005-07-20
	+	Added "no break" CSS code in outboard.php to the names display on 
		the board so that they won't break into two lines if spaces or 
		hyphens are used.

2005-05-16
	+	Fixed bug in counting rows that was throwing off the header repeat.

2005-03-07
	+	Started programming 2.2.2 release.
	+	Fixed bug in index.php that stopped the outboard window from 
		being loaded properly if the URL contained "index.php" instead
		of it being assumed by the web server.

2005-03-06
	+	Released version 2.2.1.
	+	Completed programming 2.2.1 release.
	+	Fixed install.php by removing need for register_globals = on
	+	Added a .htaccess file in root dir with a line that turns off
		register_globals in PHP.

2005-03-04
	+	Completed programming 2.2 release.
	+	Added tr entry to stylesheet.php to fix font size problem in 
		mozilla/firefox. 
	+	Added feature to allow board changes by certain users (or all)

2005-03-02
	+	Fixed 'remarks' bug in OutboardTimeclock.php which was stopping
		it from showing equal remarks on different days (details view).

2005-02-19
	+	Completed programming 2.1 release.
	+	Changed OutBoard window name to "${board_title}_outboard".
	+	Fixed $ob->getNames() bug that caused people with NULL options
		to not appear in the result list.
	+	Optimized the functions that get the first and last log dates.
	+	Added named anchors to the list of people in the OutBoard so 
		that the HTML page will reposition itself to the proper name i
		when changing dots or remarks.
	+	Added repeating headers to the Admin Editusers list of users.
	+	Added feature to open a small window (to see schedules, for
		example) if the schedule_url config option is set.

2005-02-17
	+	Completed programming 2.0 release.
	+	Removed remaining MySQL calls from most files (except DB class.)
	+	Added auto-logout of idlers (after a set amount of time.)
	+	Added Timeclock Reports feature to show logs and make timesheets.
	+	Added a get_fullname template function for programmers.
	+	Added header reprinting after a set number of people displayed

2005-02-16
	+	Updated the installation script to work with the new classes.
	+	Cleaned up the authentication code.
	+	Changed password storage in the DB to use mysql password().
	+	Changed mailto: link (in about.php) to point to my web site.

2005-02-15
    +	Started working on 2.0 release.
	+	Changed files to php extension (was php3)
	+	Moved all the database code (except for install.php) into classes.
	+	Made classes for basic authentication and configuration
	+	Moved include files into an include directory
	+	Changed the scripts to use the new classes.
	+	Updated the config.php file to work the the config class.


2002-04-11
	+	Changed the authentication system to check for Basic Auth first
	+ 	Added feature to let more than one person be admin

2001-06-12
	+	Changed IN time to be a valid MySQL date instead of 0000-00-00.
	+ 	Updated some comments in the code.

2001-06-11

outboard.php
<?php

// outboard.php
//
// This is the main PHP script that shows the OutBoard interface.
//
// 2007-02-16, richardf - 2.2.4 - added custom time for cookies
// 2006-06-15, richardf - 2.2.3 - added zebra striping option
// 2005-05-16, richardf - 2.2.2 - moved rowcount inside of if() 
// 2005-02-19, richardf - 2.1 - added named anchors to position the page on updates
// 2005-02-15, richardf - 2.0 - see CHANGES file
// 2001-06-08, richardf - 1.3 - see CHANGES file
// 2000-08-30, Richard F. Feuerriegel (richardf@acesag.auburn.edu)
// 	- Initial creation

require_once("lib/OutboardDatabase.php");
require_once("lib/OutboardAuth.php");

include_once("include/char_widths.php"); 
include_once("include/common.php"); 

// Create main objects;
$auth = new OutboardAuth();
$ob   = new OutboardDatabase();

// Set some simple variables used later in the page
$baseurl             = $_SERVER['PHP_SELF'];
$current             = getdate();
$version             = $ob->getConfig('version');
$version_date        = $ob->getConfig('version_date');
$max_visible_length  = $ob->getConfig('max_visible_length');
$cookie_time_seconds = $ob->getConfig('cookie_time_seconds');
$body_bg             = $ob->getConfig('body_bg');
$td_bg               = $ob->getConfig('td_bg');
$td_zebra1           = $ob->getConfig('td_zebra1');
$td_zebra2           = $ob->getConfig('td_zebra2');
$zebra_stripe		 = $ob->getConfig('zebra_stripe');
$td_user_bg          = $ob->getConfig('td_user_bg');
$td_text             = $ob->getConfig('td_text');
$td_lines            = $ob->getConfig('td_lines');
$link_text           = $ob->getConfig('link_text');
$windows_font_family = $ob->getConfig('windows_font_family');
$unix_font_family    = $ob->getConfig('unix_font_family');
$windows_bfs         = $ob->getConfig('windows_bfs');
$unix_bfs            = $ob->getConfig('unix_bfs');
$image_dir           = $ob->getConfig('image_dir');
$change_image        = $ob->getConfig('change_image');
$view_image          = $ob->getConfig('view_image');
$empty_image         = $ob->getConfig('empty_image');
$in_image            = $ob->getConfig('in_image');
$out_image           = $ob->getConfig('out_image');
$dot_image           = $ob->getConfig('dot_image');
$right_arrow         = $ob->getConfig('right_arrow');

// Run the installation script if the config says to
if ($ob->getConfig('installtables')) { include("include/install.php"); }


// Get the session (if there is one)
$session = $auth->getSessionCookie();

if ($ob->getConfig('authtype') == "internal") {
  $BasicAuthInUse = false;
  if ($username = getPostValue('username') and $password = getPostValue('password')) {
    $session = $ob->checkPassword($username,$password);
  }
} else {
  $BasicAuthInUse = true;
  if (! $session) {
    $username = $auth->checkBasic();
    if ($ob->isBoardMember($username)) {
      $ob->setOperatingUser($username);
      $session = $ob->setSession();
    }
  }
}

$auth->setSessionCookie($session,$cookie_time_seconds);
$username = $ob->getSession($session);

// Show the login screen if the user is not authenticated
if (! $username) { 
  $auth->setSessionCookie("",$cookie_time_seconds);
  include("include/loginscreen.php"); 
}

// if 'logout' is set, run the logout functions and go back
// to the login screen.
if (getGetValue('logout')) { 
  $ob->setSession("");
  $auth->setSessionCookie("",$cookie_time_seconds);
  include("include/loginscreen.php");
} 

if (getPostValue('exitadmin')) { 
  // trick the page into noupdate mode
  $_GET['noupdate'] = 1;
} elseif (getGetValue('adminscreen') and $ob->isAdmin() ) { 
  include("include/admin.php"); 
}

// Get the owner of the dot we want to change (might be someone else's dot)
$userid = getGetValue('userid');

// The user wants to move the dot to the Out column
if ($out = getGetValue('out')) { $ob->setDotOut($userid); }

// The user wants to move the dot to the In column
if ($in = getGetValue('in')) { $ob->setDotIn($userid); }

// The user wants to move the dot to the specified "will return by" column. The
// return variable contains the hour in the day that the user will return.
if ($return = getGetValue('return')) { $ob->setDotTime($userid,$return); }

// The user wants to change the remarks. We have to use isset() here first 
// to allow for empty remarks.
if (isset($_GET['remarks'])) { 
  $remarks = getGetValue('remarks');
  $ob->setRemarks($userid,$remarks); 
}


// Appropriately set the update flag. 
if (getGetValue('noupdate')) { 
  $update = 0; 
  if ($current['hours'] >= 6 && $current['hours'] <= 18 ) { 
    $update_msec = $ob->getConfig('reload_sec') * 1000;
  } else {
    // Set the update rate to the "night rate" if between 6:00pm and 6:00am
    $update_msec = $ob->getConfig('night_sec') * 1000;
  }
} else { 
  $update = 1; 
  $update_msec = $ob->getConfig('update_sec') * 1000;
}


?>

<HTML>
<HEAD>

<SCRIPT Language="JavaScript">
  function openWindow( window_name, url, width, height ) {
    locX = (screen.width / 2) - (width / 2);
    locY = (screen.height / 2) - (height / 2);
    window_name = window.open(url, window_name,
      "dependent=yes,resizable=yes,scrollbars=yes,screenX=" + locX 
       + ",screenY=" + locY + ",width=" + width + ",height=" + height);
    window_name.focus();
  } 
  function myReload() {
    self.location = "<?php echo $baseurl ?>?noupdate=1";
  }
  t = setTimeout("myReload()",<?php echo $update_msec ?>);
</SCRIPT>

<?php if ($launch = getGetValue('launch')) { ?>
  <Script Language="JavaScript"> window.focus(); </Script>
<?php } ?>

<TITLE>OutBoard: <?php echo $ob->getConfig('board_title') ?></TITLE>
<?php include("include/stylesheet.php"); ?>
</HEAD>
<BODY>

<Script language="JavaScript1.2">
  function change_remark(remark,userid) {
    var newremark = prompt("Enter your remarks below:",remark);
    if (newremark != null) {
      self.location="<?php echo $baseurl ?>?remarks=" 
		    + escape(newremark) + "&userid=" +userid + "#<?php echo $userid ?>";
    }
  }
</Script>

<TABLE BORDER=0 WIDTH=100%>
  <TR>
  <TD class=header><?php echo $ob->getConfig('board_title') ?></TD>
  <TD class=headernb><?php echo date($ob->getConfig('date_format')) ?></TD>
  <TD class=header align=right>
  <?php if($ob->isReadonly()) { $readonly = true; } else { $readonly = false; }?>
  <?php if(getGetValue('noupdate')) $update = 0;?>
  <?php if (! $update && ! $readonly) { ?>
    <a href="<?php echo $baseurl ?>?update=1#<?php echo $username ?>"><img src=<?php echo "$image_dir/$change_image" ?> ALT="Switch to update mode" TITLE="Switch to update mode" BORDER=0></a> 
  <?php } elseif (! $readonly) { ?>
    <a href="<?php echo $baseurl ?>?noupdate=1"><img src=<?php echo "$image_dir/$view_image" ?> ALT="Switch to view only mode" TITLE="Switch to view only mode" BORDER=0></a> 
  <?php } else { echo "&nbsp;"; } ?>
  </TD>
  </TR>
</TABLE>


<TABLE BORDER=0 WIDTH=100% ALIGN=CENTER>
<TR><TD CLASS=back>
<TABLE BORDER=0 WIDTH=100% ALIGN=CENTER>
<TR><TH></TH><TH></TH><TH colspan=10>Will return by this time:</TH><TH></TH><TH></TH></TR>

<?php $header = "<TR><TH>Name</TH><TH>In</TH>
<TH>8</TH><TH>9</TH><TH>10</TH><TH>11</TH><TH>12</TH><TH>1</TH><TH>2</TH><TH>3</TH>
<TH>4</TH><TH>5</TH><TH>Out</TH><TH>Remarks</TH></TR>";

//echo $header;
?>

<?php
// Get the latest outboard information from the database
$ob->getData();

$rowcount = 0;
$zebra = 2;
$username = urlencode($username);

while($row = $ob->getRow()) {
  $isChangeable = $ob->isChangeable($row['userid']);
  $row['userid'] = urlencode($row['userid']); 
  if (! preg_match("/\<READONLY\>/",$row['options'])) {
     $datetime = getdate($row['back']);
     if ($row['last_change'] != "") { 
       list($uname,$ip) = explode(",",$row['last_change']);
       $lastup = "Last updated by $uname from $ip on " .  $row['timestamp'] . ""; 
       $alt = "ALT=\"$lastup\" TITLE=\"$lastup\""; 
     } else {
       $alt = "";
     }
     $in = "<img src=$image_dir/$in_image $alt>"; 
     if ($datetime['year'] > $current['year']) { 
       $out = "<img src=$image_dir/$out_image $alt>"; 
       if ($update && $isChangeable) {
	     $in= "<a href=\"$baseurl?in=1&userid=".$row['userid']."#".$row['userid']."\">"
	         ."<img src=$image_dir/$empty_image BORDER=0></a>"; 
       } else {
	     $in= "<img src=$image_dir/$empty_image>"; 
       }
     } else {
       if ($update && $isChangeable) {
	     $out= "<a href=\"$baseurl?out=1&userid=".$row['userid']."#".$row['userid']."\">"
		      ."<img src=$image_dir/$empty_image BORDER=0></a>"; 
       } else {
	     $out= "<img src=$image_dir/$empty_image>"; 
       }
     }
	 for ($i = 8; $i <= 17; $i++) {
		 if ( $datetime['hours'] == $i ) { 
		   $back[$i] = "<img src=$image_dir/$dot_image $alt>"; 
		   if ($update && $isChangeable) {
			 $in= "<a href=\"$baseurl?in=1&userid=".$row['userid']."#".$row['userid']."\">"
			 ."<img src=$image_dir/$empty_image BORDER=0></a>"; 
		   } else {
			 $in= "<img src=$image_dir/$empty_image>"; 
		   }
		 } else {
		   if ($update && $isChangeable) {
			 $back[$i] = "<a href=\"$baseurl?return=$i&userid=".$row['userid']."#".$row['userid']."\">"
				."<img src=$image_dir/$empty_image BORDER=0></a>"; 
		   } else {
			 $back[$i] = "<img src=$image_dir/$empty_image>"; 
		   }
		 }
	 }
	 if ($ob->getConfig('zebra_stripe') != 0) {
		if ($rowcount % $ob->getConfig('zebra_stripe') == 0) { 
			if ($zebra == 1) { $zebra = 2; } else { $zebra = 1; }
		}
		$user_bg = "class=zebra".$zebra;
	 } else {
		$user_bg = ""; 
	 }
	 if ($row['userid'] == $username && $update && $isChangeable) { 
		 $user_bg = "class=user"; 
	 }
     if ($rowcount % $ob->getConfig('reprint_header') == 0) { echo $header; }
     echo "<TR class=norm>";
     echo "<TD WIDTH=15% $user_bg><A class=\"nobr\" name=\"".$row['userid']."\">".$row['name']."</A></TD>";
     echo "<TD WIDTH=1% $user_bg>$in</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['8']."</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['9']."</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['10']."</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['11']."</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['12']."</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['13']."</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['14']."</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['15']."</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['16']."</TD>";
     echo "<TD WIDTH=1% $user_bg>".$back['17']."</TD>";
     echo "<TD WIDTH=1% $user_bg>$out</TD>";
     if ($row['remarks'] == "") { 
       if ($update) {
	     $print_remarks = "&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "
			 ."&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "
			 ."&nbsp; &nbsp; &nbsp; &nbsp;"; 
       } else {
	     $print_remarks = "&nbsp;";
       }
     } else { 
       $visible = trim_visible($row['remarks'],$max_visible_length);
       if ($visible != $row['remarks']) {
	     $rem = $row['remarks'];
         $alt = "ALT=\"$rem\" TITLE=\"$rem\""; 
	     $print_remarks = htmlspecialchars($visible)
	                  . "<img src=$image_dir/$right_arrow BORDER=0 $alt>";
       } else {
	     $print_remarks = htmlspecialchars($visible);
       }
     }
     if ($update && $isChangeable) {
       echo "<TD WIDTH=\"55%\" $user_bg><a href=\"javascript:this.change_remark('" 
	    . addslashes(htmlspecialchars($row['remarks'])) 
	    . "','".$row['userid']."')\">$print_remarks</a></TD>";
     } else {
       echo "<TD WIDTH=\"55%\" $user_bg>$print_remarks</TD>";
     }
     echo "</TR>\n";
	 $rowcount++; 
   } // end if
} // end while

?>
</TABLE>
</TD></TR>
</TABLE>


<TABLE BORDER=0 WIDTH="100%" CELLPADDING=1 CELLSPACING=0>
  <TR>

  <?php if (! $BasicAuthInUse) { ?> 
    <TD class=small align=left width="1%">
      <a href="<?php echo $baseurl ?>?logout=1">[Logout]</a>
    </TD>
  <?php } ?>

  <?php if ($sched_url = $ob->getConfig('schedule_url')) { ?>
    <TD class=small align=center>
     <a href="javascript:void(0)" 
	onClick="openWindow('scheduleWindow','<?php echo $sched_url ?>',550,600)"><?php echo $ob->getConfig('schedule_name'); ?></a>
    </TD>
  <?php } ?>

  <TD class=small align=center>
  <?php include("include/about.php"); ?>
  </TD>

  <TD class=small align=left width="1%">
      <a href="timeclock.php" target=_blank>[Report]</a>
  </TD>

  <?php if ($ob->isAdmin()) { ?>
  <TD class=small align=left width="1%">
      <a href="<?php echo $baseurl ?>?adminscreen=1">[Admin]</a>
  </TD>
  <?php } ?>

  </TR>
</TABLE>

</BODY>
</HTML>
